src/vs/workbench/contrib/chat/browser/chatViewTitle.ts

```diff
@@ -0,0 +1,178 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import './chatViewTitle.css';
+
+import { h, append } from '../../../../base/browser/dom.js';
+import { Disposable } from '../../../../base/common/lifecycle.js';
+import { localize } from '../../../../nls.js';
+import { IConfigurationService } from '../../../../platform/configuration/common/configuration.js';
+import { IViewDescriptorService, IViewContainerModel } from '../../../common/views.js';
+import { ActivityBarPosition, LayoutSettings } from '../../../services/layout/browser/layoutService.js';
+import { IChatModel } from '../common/chatModel.js';
+
+export interface IChatViewTitleDelegate {
+	getCurrentModel(): IChatModel | undefined;
+	updatePrimaryTitle(title: string): void;
+	requestLayout(): void;
+}
+
+export class ChatViewTitleController extends Disposable {
+	private readonly viewContainerModel: IViewContainerModel | undefined;
+	private hasMultipleVisibleViews = false;
+	private currentPrimaryTitle: string;
+	private _secondaryTitleContainer: HTMLElement | undefined;
+	private _secondaryTitle: HTMLElement | undefined;
+
+	constructor(
+		private readonly viewId: string,
+		private readonly delegate: IChatViewTitleDelegate,
+		@IConfigurationService private readonly configurationService: IConfigurationService,
+		@IViewDescriptorService private readonly viewDescriptorService: IViewDescriptorService,
+	) {
+		super();
+		const viewContainer = this.viewDescriptorService.getViewContainerByViewId(this.viewId);
+		if (viewContainer) {
+			this.viewContainerModel = this.viewDescriptorService.getViewContainerModel(viewContainer);
+			this.hasMultipleVisibleViews = this.viewContainerModel.visibleViewDescriptors.length > 1;
+			this._register(this.viewContainerModel.onDidAddVisibleViewDescriptors(() => this.handleVisibleViewDescriptorsChanged()));
```

Comment:
I think it would be easier to follow the logic if instead of having different handlers for the different specific things that can change, you would just have one `update` method which always runs when anything changes, checks the state of everything, and renders the content as needed. Caching something like `hasMultipleVisibleViews` is premature optimization IMO